extension/encoding/awslogsencodingextension/internal/unmarshaler/s3-access-log/unmarshaler.go

```diff
@@ -0,0 +1,238 @@
+	}
+}
+
+type resourceAttributes struct {
+	bucketOwner string
+	bucketName  string
+}
+
+func (s *s3AccessLogUnmarshaler) UnmarshalLogs(buf []byte) (plog.Logs, error) {
+	scanner := bufio.NewScanner(bytes.NewReader(buf))
+
+	logs, resourceLogs, scopeLogs := s.createLogs()
+	resourceAttr := &resourceAttributes{}
+	for scanner.Scan() {
+		log := scanner.Text()
+		if err := handleLog(resourceAttr, scopeLogs, log); err != nil {
+			return plog.Logs{}, err
+		}
+	}
+
+	if err := scanner.Err(); err != nil {
+		return plog.Logs{}, fmt.Errorf("error reading log line: %w", err)
+	}
+
+	s.setResourceAttributes(resourceAttr, resourceLogs)
+	return logs, nil
+}
+
+// createLogs with the expected fields for the scope logs
+func (s *s3AccessLogUnmarshaler) createLogs() (plog.Logs, plog.ResourceLogs, plog.ScopeLogs) {
+	logs := plog.NewLogs()
+	resourceLogs := logs.ResourceLogs().AppendEmpty()
+	scopeLogs := resourceLogs.ScopeLogs().AppendEmpty()
+	scopeLogs.Scope().SetName(metadata.ScopeName)
+	scopeLogs.Scope().SetVersion(s.buildInfo.Version)
+	return logs, resourceLogs, scopeLogs
+}
+
+// setResourceAttributes based on the resourceAttributes
+func (s *s3AccessLogUnmarshaler) setResourceAttributes(r *resourceAttributes, logs plog.ResourceLogs) {
+	attr := logs.Resource().Attributes()
+	attr.PutStr(semconv.AttributeCloudProvider, semconv.AttributeCloudProviderAWS)
+	if r.bucketName != "" {
+		attr.PutStr(semconv.AttributeAWSS3Bucket, r.bucketName)
+	}
+	if r.bucketOwner != "" {
+		attr.PutStr(attributeAWSS3BucketOwner, r.bucketOwner)
+	}
+}
+
+// getFullValue checks if value starts with a quote. If it does,
+// then it looks for the rest of the value until the end quote.
+// Otherwise, it returns the value as it is.
+func getFullValue(value string, remaining string) (string, string, error) {
+	if len(value) == 0 || value[0] != '"' {
+		return value, remaining, nil
+	}
+
+	value = value[1:] // remove first quote
+	if len(value) > 0 && value[len(value)-1] == '"' {
+		// remove last quote from the value
+		return value[:len(value)-1], remaining, nil
+	}
+
+	// value ends on next quote, get the rest
+	end := strings.IndexByte(remaining, '"')
+	if end == -1 {
+		return "", "", fmt.Errorf("value %q has no end quote", value)
+	}
+
+	value = value + " " + remaining[:end]
+	remaining = remaining[end+1:]
+	if len(remaining) > 0 && remaining[0] == ' ' { // remove next space if it exists
+		remaining = remaining[1:]
+	}
+	return value, remaining, nil
+}
+
+func handleLog(resourceAttr *resourceAttributes, scopeLogs plog.ScopeLogs, log string) error {
+	record := plog.NewLogRecord()
+
+	remaining := log
+	var i int
+	var value string
+	var err error
+	for i = 0; remaining != ""; i++ {
+		if i >= len(attributeNames) {
+			return errors.New("values in log line exceed the number of available fields")
+		}
+
+		value, remaining, _ = strings.Cut(remaining, " ")
+		if value, remaining, err = getFullValue(value, remaining); err != nil {
+			return err
+		}
```

Comment:
I suggest replacing both of these with a single function, like

```go
value, remaining, err := scanField(remaining)
if err != nil {
	return err
}
```

`scanField` would look something like:

```go
func scanField(s string) (field string, remaining string, err error) {
	if s == "" {
		return "", "", io.EOF
	}
	if s[0] != '"' {
		field, remaining, _ = strings.Cut(s, " ")
		return field, remaining, nil
	}
	field, remaining, ok := strings.Cut(s[1:], `"`)
	if !ok {
		return "", "", errors.New("unmatched quote")
	}
	return field, remaining, nil
}
```

This assumes there's no possibility of `"` showing up as an escaped character or something like that.

Reply:
I have done this change. The function needed to be adjusted, because it does not consider the next space after `"`.